### DISCUSSION TITLE
Confusion about cache of `grad(my_jitted_function)`

### DISCUSSION THREAD

**Discussion by langmore** [jax-ml/jax#9590] [Q&A] [answered]
Given a jitted function `f`, I can check cache size with `f._cache_size()`.  Now, given `df = grad(f)`, I assume calling `df(x)` populates cache *somewhere*. Where and how do I check its size?

```python
@jax.jit
def f(x):
  return x**2
df = jax.grad(f)

df(1.)  # Hopefully this populates some cache.
print(f._cache_size())
==> 0  # Was expecting 1

f(1.)  # Of course this populates cache
print(f._cache_size())
==> 1  # As expected, directly calling f populates cache
```

**Comment by mattjj** [comment, marked as answer]
Thanks for the question!

It's a bit complicated, both for fundamental reasons (caching transformed versions of functions given the way JAX's tracing works) and for transient historical-path-dependent reasons.

There are two global jit caches: one for C++ dispatch and the other for dispatching handled by Python. The Python cache is fully general in the sense that it works with all manner of transformations applied to the jitted function, while the C++ cache only works with buffers-in-buffers-out dispatching, and hence not transformation-of-jit cases. (The C++ cache is populated by stealing entries from the Python cache when it can.)

The `_cache_size()` API (which isn't public AFAIK, is it even documented?) in your example queries the C++ cache only. But because that cache doesn't support the grad-of-jit case, you won't see it increase when you evaluate grad-of-jit calls, as in your example.

To see the cache grow in grad-of-jit calls, you need to dig pretty carefully:

```python
import jax

@jax.jit
def f(x):
  return x**2
df = jax.grad(f)

def hilariously_tricky_cache_size():
  from jax._src import dispatch
  return sum(len(entries) for entries in dispatch._xla_callable.__closure__[1].cell_contents.values())

print(hilariously_tricky_cache_size())  # 0
df(1.)  # Hopefully this populates some cache.
print(hilariously_tricky_cache_size())  # 2
```

The reason there are two entries is that there's a cached jitted forward pass and a cached jitted backward pass.

If you'd like a public API for that, open a feature request and let's discuss it!

WDYT?

**Reply by langmore** [reply]
Thanks. I have a memory leak that looks a lot like cache growing. Turns out it was. Your suggested cache-counting-call verified this. Thanks!  Filed https://github.com/google/jax/issues/9600

**Reply by patrick-kidger** [reply]
FWIW I've ran into such "memory leaks" before, in particular when running tests on memory-limited machines (Arguable if they're leaks; the cache is working as intended.) If it's helpful -- I use the following snippet (ran between every test) to clear out all/most of the caches that JAX uses every process time memory grows beyond 4GB.

```python
def clear_caches():
    process = psutil.Process()
    if process.memory_info().vms > 4 * 2**30:  # >4GB memory usage
        for module_name, module in sys.modules.items():
            if module_name.startswith("jax"):
                for obj_name in dir(module):
                    obj = getattr(module, obj_name)
                    if hasattr(obj, "cache_clear"):
                        obj.cache_clear()
        gc.collect()
```

**Reply by langmore** [reply]
I see that `xla_callable` is now a "public" member of `dispatch`, e.g. I can use `dispatch.xla_callable`.  @mattjj  Is your above answer still a good way to query the Python cache?

**Reply by VolodyaCO** [reply]
I think the new 0.4.4 release broke `hilariously_tricky_cache_size`.

**Reply by yashk2810** [reply]
With jax 0.4.4, jit and pjit's implementation were merged so the usage of the above internal APIs won't work since those codepaths are  not being executed anymore.

But we will probably expose a public API to do this. You can downgrade to 0.4.3 if you are dependent on this behavior for now.

**Reply by VolodyaCO** [reply]
Is there an alternative to get the cache size for the time being?

**Reply by yashk2810** [reply]
Why are you depending on the cache size though?

**Reply by VolodyaCO** [reply]
I am not. I am just testing on my side that `jax.clear_backends()` works as expected. I need to clear my cache from time to time to avoid out of memory errors. I use the `hilariously_tricky_cache_size` to check that I don't have anything in cache at all.

**Reply by yashk2810** [reply]
I think it does work as expected. We have tests for it to make sure it does! The problem is that you are accessing internal APIs here which will break as it did with this release. If you want you can look at `pjit._pjit_lower_cached.cache_info` but this can break again in the future (if we decide to change the name or something else). 

That being said, we will work on a public API.

**Reply by VolodyaCO** [reply]
Is there any progress here?

**Reply by patrick-kidger** [reply]
The only solution I've ever found to this is to restart the Python process.

For example I found that the tests for Lineax (run all at once via `pytest`) ended up OOM'ing. The solution was to run the tests via `python -m tests`, which triggers [this file](https://github.com/google/lineax/blob/main/tests/__main__.py), which runs pytest separately on each file.

**Reply by VolodyaCO** [reply]
oh wow. but is there a way right now in the public API to know what the size of the cache is? or is this still something that needs to be done?
